zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -0,0 +1,612 @@
+    tx_value: Word<F>,
+    tx_call_data_length: Cell<F>,
+    tx_call_data_gas_cost: Cell<F>,
+    rw_counter_end_of_reversion: Cell<F>,
+    is_persistent: Cell<F>,
+    sufficient_gas_left: RangeCheckGadget<F, MAX_GAS_SIZE_IN_BYTES>,
+    transfer_with_gas_fee: TransferWithGasFeeGadget<F>,
+    code_hash: Cell<F>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for BeginTxGadget<F> {
+    const NAME: &'static str = "BeginTx";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::BeginTx;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        // Use rw_counter of the step which triggers next call as its call_id.
+        let call_id = cb.curr.state.rw_counter.clone();
+
+        let [tx_id, rw_counter_end_of_reversion, is_persistent] = [
+            CallContextFieldTag::TxId,
+            CallContextFieldTag::RwCounterEndOfReversion,
+            CallContextFieldTag::IsPersistent,
+        ]
+        .map(|field_tag| cb.call_context(Some(call_id.expr()), field_tag));
+
+        let [tx_nonce, tx_gas, tx_caller_address, tx_callee_address, tx_is_create, tx_call_data_length, tx_call_data_gas_cost] =
+            [
+                TxContextFieldTag::Nonce,
+                TxContextFieldTag::Gas,
+                TxContextFieldTag::CallerAddress,
+                TxContextFieldTag::CalleeAddress,
+                TxContextFieldTag::IsCreate,
+                TxContextFieldTag::CallDataLength,
+                TxContextFieldTag::CallDataGasCost,
+            ]
+            .map(|field_tag| cb.tx_context(tx_id.expr(), field_tag));
+        let [tx_gas_price, tx_value] =
+            [TxContextFieldTag::GasPrice, TxContextFieldTag::Value].map(
+                |field_tag| cb.tx_context_as_word(tx_id.expr(), field_tag),
+            );
+
+        // Add first step constraint to have both rw_counter and tx_id to be 1
+        cb.add_constraint_first_step(
+            "rw_counter is initialized to be 1",
+            1.expr() - cb.curr.state.rw_counter.expr(),
+        );
+        cb.add_constraint_first_step(
+            "tx_id is initialized to be 1",
+            1.expr() - tx_id.expr(),
+        );
+
+        // Increase caller's nonce.
+        // (tx caller's nonce always increases even tx ends with error)
+        cb.account_write(
+            tx_caller_address.expr(),
+            AccountFieldTag::Nonce,
+            tx_nonce.expr() + 1.expr(),
+            tx_nonce.expr(),
+        );
+
+        // TODO: Implement EIP 1559 (currently it only supports legacy
+        // transaction format)
+        // Calculate transaction gas fee
+        let mul_gas_fee_by_gas = MulWordByU64Gadget::construct(
+            cb,
+            tx_gas_price.clone(),
+            tx_gas.clone(),
+            true,
+        );
+
+        // TODO: Take gas cost of access list (EIP 2930) into consideration.
+        // Use intrinsic gas
+        let intrinsic_gas_cost =
+            select::expr(tx_is_create.expr(), 53000.expr(), 21000.expr())
```

Comment:
Should we move these numbers to a `const`. Will be easier to parametrize I think.

Reply:
Fixed in [`bb2f101`](https://github.com/appliedzkp/zkevm-circuits/pull/230/commits/bb2f1012ec8dd8b6d5f91f87bba2757180fdef60).